### PR TITLE
[Feat]: Specdec Streaming: RDMA + Multinode

### PR DESCRIPTION
**Pull request by h-guo18**
### What does this PR do?

Type of change: New feature

Multi-node **streaming** training for speculative decoding (EAGLE3 / DFlash):
a live `vllm serve` captures the target model's hidden states and moves them
straight to the trainer over **NIXL RDMA** — no disk round-trip. The streaming
dataset is map-style — each rank fetches only its own `DistributedSampler` shard
(concurrency from `dataloader_num_workers`), round-robins across multiple serve
replicas (`server_urls`), and scales to multi-node DDP. Serve-side tensor
parallelism (TP>1) is supported: hidden states are replicated across TP ranks, so
rank 0 alone owns the pool + transfer.

### How

- `RdmaHiddenStatesConnector` — out-of-tree vLLM connector (no vLLM source edits):
  one pre-registered pinned NIXL pool per serve, a ring slot per request, and a
  small HTTP sidecar serving transfer metadata. The trainer RDMA-READs the slot
  into a per-worker buffer. RDMA is the **only** transport (the earlier
  disk/safetensors path is removed).
- Map-style dataset + multi-node accelerate launch (`--machine_rank`, optional
  Slurm `--segment` to keep nodes in one NVLink domain).

### Usage

```yaml
data:
  mode: streaming
  streaming_server_url: "http://node0:8000,http://node1:8000"  # round-robin
```

### Validation (Qwen3-8B, oci-nrt H100)
sandbox CI: https://gitlab-master.nvidia.com/omniml/integration/nmm-sandbox/-/jobs/337489812

**1. End-to-end convergence — EAGLE3 & DFlash, 5000 steps.** Both algorithms
converge and export a deployable draft; the DFlash drafts also serve under vLLM
speculative decoding (8/8 smoke prompts pass).

| algorithm | topology (nodes) | train loss (step 0 → 5000) | vLLM draft acc-len |
|---|---|---|---|
| EAGLE3 | 2 serve TP=2 + 2 trainer DDP (4) | 37.1 → 8.20 | — |
| DFlash | 1 serve TP=1 + 1 trainer (2)      | 11.7 → 5.56 | 1.11 |
| DFlash | 2 serve TP=2 + 2 trainer DDP (4)  | 10.9 → 5.26 | 1.19 |

**2. Scalability — 1 → 12 nodes (EAGLE3, 200 steps).** Throughput scales ~23× across the
sweep below. The step-time growth is cross-node DDP all-reduce, not the streaming path —
RDMA (~0.33 ms/req @ 2 MB, ~47 GB/s host-pinned READ) is never the bottleneck. Scale
serve + trainer nodes together for near-linear speedup.

| serve / trainer | nodes | step time | samples / step | samples / sec (global) | acc @ step 200 |
|---|---|---|---|---|---|
| 1 serve / 1 rank (co-located, 1 node 2 GPU) | 1 | 0.23 s | 1  | 4.4 | [0.141, 0.094, 0.072] |
| 1 serve / 1 rank (cross-node)               | 2 | 0.23 s | 1  | 4.3 | [0.137, 0.105, 0.074] |
| 2 serve / 8 ranks                           | 3 | 0.26 s | 8  | 31.1 | [0.215, 0.126, 0.097] |
| 4 serve / 16 ranks (2 trainer nodes)        | 6 | 0.28 s | 16 | 56.5 | [0.217, 0.148, 0.110] |
| 8 serve / 32 ranks (4 trainer nodes)        | 12 | 0.31 s | 32 | 101.9 | [0.235, 0.165, 0.137] |

**3. Serve-side TP correctness.** TP=1 vs TP=2 draft top-1 accuracy track
step-for-step (hidden states are replicated across TP ranks).

serve-tp-acc

### Before your PR is "*Ready for review*"

- Backward compatible?: ❌ — streaming is now RDMA-only; `server_url` → `server_urls`;
  the disk transport (`HS_TRANSPORT`, `streaming_shared_storage_path`) is removed.
- New tests?: ✅ `tests/unit/torch/speculative/plugins/test_hf_streaming_dataset.py`
  (map-style dataset + mocked RDMA fetch).
- Updated Changelog?: ❌
- Claude approval?: ❌

### CODE REVIEW

File: modelopt/torch/speculative/plugins/hf_training_args.py [hunk 1]
```diff
@@ -62,7 +64,6 @@ class DataArguments(BaseModel):
     sample_size: int = -1
     streaming_server_url: str | None = None
     streaming_model_name: str | None = None
-    streaming_prefetch: int = Field(default=64, ge=1)
     # Mirror of the vLLM connector's ``shared_storage_path``; trainer-side allowlist.
     streaming_shared_storage_path: str | None = None
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_training_args.py]: Intentional — we are dropping `streaming_prefetch` entirely and do not need backward compatibility. Any old recipe/yaml still referencing it should be updated to remove the field. Resolving.

File: modelopt/torch/speculative/plugins/hf_streaming_dataset.py [hunk 2]
```diff
@@ -467,13 +365,38 @@ def __init__(
...
+        round-robin cursor over ``server_urls``), keyed by PID.
+        """
+        pid = os.getpid()
+        if getattr(self, "_client_pid", None) != pid:
+            self._http = httpx.Client(
+                timeout=httpx.Timeout(self.config.request_timeout, connect=10.0)
+            )
+            self._client_pid = pid
+            self._rr = 0
+        return self._http
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply modelopt/torch/speculative/plugins/hf_streaming_dataset.py]: Good catch — fixed. `_client()` now seeds the round-robin cursor at `rank * num_workers + worker_id` (rank from the `RANK` env var, worker info from `torch.utils.data.get_worker_info()`) instead of `0`, so cold-start first-fetches fan out across replicas instead of all pinning `server_urls[0]`. Rank is factored in as you suggested, so rank-N worker-0 no longer collides with rank-(N+1) worker-0. Landing in the next push.

File: tools/launcher/common/eagle3/train_eagle_streaming.sh [hunk 3]
```diff
@@ -218,36 +202,58 @@ wait_vllm_ready() {
...
+        echo "machine_rank=${mrank}: training done, skipping export (head trainer handles it)."
+        return 0
+    fi
+
+    # Derive checkpoint dir from the forwarded training.output_dir= dotlist (EAGLE default)
+    # so EAGLE/DFlash runs each export their own dir. EXPORT_EXTRA_ARGS lets DFlash on a
+    # custom-modeling base (e.g. Kimi) pass --trust_remote_code; empty by default.
+    local out_dir
+    out_dir=$(printf '%s\n' "${SCRIPT_ARGS[@]}" | sed -n 's/^training\.output_dir=//p' | tail -1)
+    out_dir="${out_dir:-/scratchspace/eagle3}"
```
[the remark this answers is not in the document]
Comment (h-guo18) [reply tools/launcher/common/eagle3/train_eagle_streaming.sh]: Good point — replaced the hard-coded `/scratchspace/eagle3` fallback with a fail-loud guard: if no `training.output_dir=` was forwarded in `SCRIPT_ARGS`, the export step now errors out (`return 1`) instead of silently exporting from the wrong dir. Every streaming yaml already forwards it, so this only fires on a misconfigured run. Landing in the next push.

**Comment by ChenhanYu** [comment]
Need CHANGELOG and example README update.

**Review by ChenhanYu — APPROVED** [review]
